### PR TITLE
WIP: OPMPHM, hsearch and unordered_map benchmark

### PR DESCRIPTION
**Pull request by KurtMi**
Benchmark suite for the comparison between OPMPHM and hsearch, unordered_map.
Plus there is something to read in the abgaben repo.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -97,6 +97,10 @@ if (BUILD_FULL OR BUILD_STATIC OR BUILD_SHARED)
 	add_subdirectory (examples)
 	add_subdirectory (benchmarks)
 
+	if (BUILD_BENCHMARKS_KSLOOKUP)
```
Comment (markus2330) [inline CMakeLists.txt]: its ok to include them always. The variability of the build system should be kept as simple as possible.

Comment (markus2330) [inline CMakeLists.txt]: and the adding should happen within benchmarks/CMakeLists.txt

File: benchmarks/kslookup/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
+include (LibAddMacros)
```
Comment (markus2330) [inline benchmarks/kslookup/CMakeLists.txt]: is it necessary to have another CMakeLists.txt for this benchmark?

File: benchmarks/kslookup/benchmark.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+#define GENDATA_KEY_VALUE "some data"
+#define EXPORT_PLUGIN "dump"
+
+//rand
+void initRand (void);
+unsigned int genRand (int modul);
+
+//stat helper
+int median (int values[], int count);
```
Comment (markus2330) [inline benchmarks/kslookup/benchmark.h]: seems to be useful for other benchmarks, too.

**Comment by markus2330** [comment]
Thank you! I will look at it at a later time.

File: benchmarks/kslookup/gendata.c [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+{
+	initRand ();
+	//plugin open
+	KeySet * modules = ksNew(0, KS_END);
+	elektraModulesInit(modules, 0);
+	KeySet * conf = ksNew (0, KS_END);
+	Plugin * plugin = elektraPluginOpen(EXPORT_PLUGIN, modules, conf, 0);
+	if(!plugin)
+	{
+		fprintf (stderr, "dump plugin could not be opened\n");
```
Comment (markus2330) [inline benchmarks/kslookup/gendata.c]: might not be dump (use EXPORT_PLUGIN)

File: benchmarks/kslookup/gendata.c [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+		//fill-up
+		int j = 0;
+		for (int i = 0;i < indexOfSlash[depth] - 1;++i)
+		{
+			if (indexOfSlash[j] == i)
+			{
+				++j;
+				randomString[i] = '/';
+			}else
```
Comment (markus2330) [inline benchmarks/kslookup/gendata.c]: should be `} else {`

File: benchmarks/kslookup/gendata.c [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		randomString[indexOfSlash[depth] - 1] = '\0';
+
+		//check unique
+		Key * key_found = ksLookupByName(checkIfUnique, randomString, KDB_O_NONE);
+		if (key_found != NULL)
+		{
+			unique = false;
+		}
+		if(!unique) free (randomString);
+	}
```
Comment (markus2330) [inline benchmarks/kslookup/gendata.c]: `} while()`

File: benchmarks/kslookup/gendata.c [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+		char * name = generateKeyName (ks);
+		ksAppendKey (ks, keyNew (name, KEY_VALUE, GENDATA_KEY_VALUE, KEY_END));
+		free (name);
+	}
+
+	return ks;
+}
+
+/* Generates a random String, not present as KeyName in
+ * the given KeySet.
```
Comment (markus2330) [inline benchmarks/kslookup/gendata.c]: which kind of random string? How are `/` introduced?

File: benchmarks/kslookup/gendata.h [hunk 8]
```diff
@@ -0,0 +1,15 @@
+#ifndef BENCHMARKS_KSLOOKUP_GENDATA_H
+#define BENCHMARKS_KSLOOKUP_GENDATA_H
+
+#define ALPHABET "abcedfghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Comment (markus2330) [inline benchmarks/kslookup/gendata.h]: why not use `isalnum()`?

Comment (markus2330) [inline benchmarks/kslookup/gendata.h]: for benchmark `setlocale(LC_ALL, "C")` is no problem

File: benchmarks/kslookup/hsearch.c [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+	elektraPluginClose (plugin, 0);
+	elektraModulesClose(modules, 0);
+	ksDel (modules);
+
+	return out;
+}
+
+/* generates the ENTRYs for the hash map
+ */
+ENTRY * prepareData (KeySet * ks)
```
Comment (markus2330) [inline benchmarks/kslookup/hsearch.c]: Where is ENTRY defined? ENTRY seems to be a too general name. And why is it a macro?

File: benchmarks/kslookup/kslookup.c [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	for (int v=1;v <= KEYSET_VERSIONS;++v)
+	{
+		char filename_out [BUFFER_FILENAME];
+		char c_mode;
+		if (mode)
+			c_mode = 'b';
+		else
+			c_mode = 's';
+		sprintf (&filename_out[0], "kslookup_%c_%i.bench",c_mode,v);
+		FILE * output = fopen (&filename_out[0], "w");
```
Comment (markus2330) [inline benchmarks/kslookup/kslookup.c]: Why not simply `filename_out`?

File: benchmarks/kslookup/kslookup.c [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+	return 1;
+}
+
+/* The actual benchmark procedure for the search, executed for each
+ * setting.
+ */
+int search (int n, int r, KeySet * data, int * times)
+{
+	Key * keysearchfor = keyNew(0);
```
Comment (markus2330) [inline benchmarks/kslookup/kslookup.c]: Isn't that a compiler warning? Should be terminated with `KEY_END`.

File: benchmarks/kslookup/kslookup.c [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	Key * keyfound;
+	struct timeval start;
+	struct timeval end;
+	//this array saves all the times to search each key
+	int keys_searched_for[n];
```
Comment (markus2330) [inline benchmarks/kslookup/kslookup.c]: Name does not really state what it is for.
Use camelCase.

File: benchmarks/kslookup/kslookup.c [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+		keySetName (keysearchfor, keyName (data->array[search_for]));
+
+		gettimeofday (&start, 0);
+//MEASURE START
+
+		//TODO KURT set hash search flag
+		keyfound = ksLookup(data, keysearchfor, KDB_O_NONE);
+//MEASURE END
+		gettimeofday (&end, 0);
+		keys_searched_for[search_for] = (int) (end.tv_sec - start.tv_sec) * 1000000 +
```
Comment (markus2330) [inline benchmarks/kslookup/kslookup.c]: avoid magic number 1000000

File: benchmarks/kslookup/kslookup.c [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+{
+	KeySet * out = ksNew(size, KS_END);
+	//plugin open
+	KeySet * modules = ksNew(0, KS_END);
+	elektraModulesInit(modules, 0);
+	KeySet *conf = ksNew (0, KS_END);
+	Plugin * plugin = elektraPluginOpen(EXPORT_PLUGIN, modules, conf, 0);
+	if(!plugin)
+	{
+		fprintf (stderr, "dump plugin could not be opened\n");
```
Comment (markus2330) [inline benchmarks/kslookup/kslookup.c]: again it might not be dump

File: benchmarks/kslookup/unorderedmap.cpp [hunk 15]
```diff
@@ -0,0 +1,174 @@
+#include <iostream>
+#include <cstdlib>
+#include <fstream>
+#include <sstream>
+
+#include <modules.hpp>
+
+#include "unorderedmap_Search.hpp"
```
Comment (markus2330) [inline benchmarks/kslookup/unorderedmap.cpp]: should be lowercase without underscores (and shorter).

File: benchmarks/kslookup/unorderedmap.cpp [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+			output.close ();
+			delete ks;
+		}
+	}
+}
+
+kdb::KeySet * readKeySet (int size, int version)
+{
+	kdb::KeySet * out = new kdb::KeySet (size, KS_END);
```
Comment (markus2330) [inline benchmarks/kslookup/unorderedmap.cpp]: no new needed here, simply use
`kdb::KeySet out(size, KS_END);`

File: benchmarks/kslookup/unorderedmap.cpp [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+				output << median (times, REPEATS) << std::endl;
+			}
+
+			output.close ();
+			delete ks;
+		}
+	}
+}
+
+kdb::KeySet * readKeySet (int size, int version)
```
Comment (markus2330) [inline benchmarks/kslookup/unorderedmap.cpp]: return per value

File: benchmarks/kslookup/hsearch.c [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+{
+	ENTRY * ep;
+	struct timeval start;
+	struct timeval end;
+
+	gettimeofday (&start, 0);
+//MEASURE START
+
+	//create
+	hcreate (k);
```
Comment (markus2330) [inline benchmarks/kslookup/hsearch.c]: why not separate create and insert time measurements?

File: benchmarks/kslookup/hsearch.c [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+			return -1;
+		}
+	}
+//MEASURE END
+	gettimeofday (&end, 0);
+	times[r] = (int) (end.tv_sec - start.tv_sec) * 1000000 +
+						(end.tv_usec - start.tv_usec);
+
+	//cleanup
+	hdestroy ();
```
Comment (markus2330) [inline benchmarks/kslookup/hsearch.c]: why no benchmark for destroy?

File: benchmarks/kslookup/unorderedmap.cpp [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+			double diff_k = MAX_BUCKET_STEP - MIN_BUCKET_STEP;
+			double ratio = diff_k/diff_n;
+			int bucket_step_count = static_cast<int> ( (n*ratio) + MIN_BUCKET_STEP );
+			if(bucket_step_count > MAX_BUCKET_STEP)
+				bucket_step_count = MAX_BUCKET_STEP;
+			int bucket_step = static_cast<int> ( n/bucket_step_count );
+			if (bucket_step * bucket_step_count != n)
+				++bucket_step;
+
+			for (int k = n; k < n*3;k+=bucket_step)
```
Comment (markus2330) [inline benchmarks/kslookup/unorderedmap.cpp]: magic number 3

**Comment by markus2330** [comment]
Looks good for a benchmark suite. The main problem is that its not well integrated with already existing benchmarks. Additionally, the cpp benchmark should be in src/bindings/cpp/benchmarks/. Its ok if you refactor existing benchmarks, maybe with discussion beforehand for larger changes.

**Comment by markus2330** [comment]
Do you plan to continue on this pull request or can we close it?

**Comment by KurtMi** [comment]
will be continued!

**Comment by KurtMi** [comment]
Pull request will be closed, due to inactivity from my side.

**Pull request closed without merging**